### ISSUE TITLE
bin/elasticsearch is not setting pid in provided pidfile

### ISSUE THREAD

**Issue by meash-nrel** [elastic/elasticsearch#25385]
**Elasticsearch version**: 5.4.1

**Plugins installed**: []

**JVM version** (`java -version`):
java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)

**OS version** (`uname -a` if on a Unix-like system):
Linux xxxxxx 2.6.32-573.18.1.el6.x86_64 #1 SMP Tue Feb 9 22:46:17 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:
When running elasticsearch 5.4.1 as a service, via the installed init.d file, it is properly creating and passing in the pidfile to _/bin/elasticsearch_ via its -p parameter... however _/bin/elasticsearch_ is never writing the process id to that provided pidfile.  This lack of writing pid causes the init.d system to never see that the service is properly running, so you can not issue a `service elasticsearch stop` command. Another side effect is that you if you run `service elasticsearch start` multiple times, each creates a separate running instance (then causing whatever conflicts arise from having the same named node running multiple times).  Any 5.4.1 node process must be manually killed as service stop doesn't work.

This is not a permission issue. The PID_DIR (the default of /var/run/elasticsearch) is properly created and assigned to elasticsearch by the init.d file.  I have an 1.7.6 instance running on this machine and it is writing to its own pidfile as expected.  (FYI There is no conflict between my new 5.4.1 node and this 1.7.6 node, as the legacy node is entirely self-contained within /opt/elasticsearch/elasticsearch-1.7.6-legacy, and writes pid to a separate directory, whereas 5.4.1 was RPM installed and uses all default dirs.  1.7.6 used to be RPM installed as well, and was writing pid to default dir /var/run/elasticsearch just fine until it was moved to its own isolated dirs.)    

init.d is working as written and passing pidfile, and elasticsearch can write to that file, so it seems that _/bin/elasticsearch_ itself is not performing the write.

Bottom line: /bin/elasticsearch isn't writing the pid.  Dir and file are owned by elasticsearch user and are RW so it is not a permission issue (and as confirmation, this exact same pid path/file worked on a prior ES version).

**Steps to reproduce**:

1. Installed latest elasticsearch 5.4.x via RPM from Elastic's [instructions](https://www.elastic.co/guide/en/elasticsearch/reference/current/rpm.html)
2. Run installed init.d service: `service elasticsearch start`
3. Try to stop service, nothing happens:  `service elasticsearch stop`
4. Can run init.d service again (multiple instances get started): `service elasticsearch start`

**Comment by meash-nrel** [comment]
For now i have worked around this issue in the init.d file, with this line right before the end of start() function:

`sleep 10 && pgrep -f $prog | sed '$!d' > $pidfile`

**Comment by nik9000** [comment]
I just tested in a fresh VM and it wrote the pidfile just fine. I didn't try and make a 1.7.6 node running in opt.

I wonder if you got anything in the logs when the 5.6.1 node started up that might hint what is up.

**Comment by meash-nrel** [comment]
Nothing at all unusual in the log file starting up - bootstrap check ok and node starts as normal.

Some odd behavior - On a whim, I moved PID_DIR to CONF_DIR in /etc/init.d/elasticsearch.  I started the service in init.d and it worked as intended (wrote the pid) -- but only the one time.  I stopped the service & started it again through _service_ calls, changing nothing in init.d, and it failed on the subsequent (and every other) attempt -- did not write the pid.  Happened with or without me removing the 0-byte pid file that init.d creates.

Can you try repeatedly calling _service elasticsearch start_ and _stop_?

**Comment by jasontedor** [comment]
This does not reproduce for me, even after repeatedly calling `service elasticsearch start` and `service elasticsearch stop`

> however */bin/elasticsearch* is never writing the process id to that provided pidfile

It's not the startup script that writes the PID file, it's Elasticsearch itself. And if Elasticsearch fails to write the PID file, we abort startup. Is something in your environment killing this file after it's written?

You can `strace` startup to see the write:

```bash
$ strace -e open,write -f bin/elasticsearch -p /tmp/es.pid
[pid 13243] open("/tmp/es.pid", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 45
[pid 13243] write(45, "13205", 5)       = 5
```

Again, if *any* portion of writing this file fails, we abort startup. So if your process is starting successfully (are you sure that it is?) then this PID file is correctly being written. If it's not there and the process is running, some external force on your system is removing it.

I'm closing this as not a bug. If you need additional help debugging, please ask on the [forum](https://discuss.elastic.co), we reserve GitHub for verified bug reports and feature requests. I'm happy to reopen this is a bug is verified here.
